File: parser/transaction.go
package parser

import (
	"errors"
	"fmt"
	"regexp"
	"strings"
	"unicode"
	"unicode/utf8"

	"github.com/thaodangspace/money-bot/domain"
)

const (
	MaxInputRunes    = 2000
	MaxCategoryRunes = 120
	MaxNoteRunes     = 500
)

var (
	ErrNoTransaction = errors.New("transaction not recognized")
	transactionRe    = regexp.MustCompile(`^(.+?)\s+([0-9][0-9.,]*\s*(?:k|tr|m|đ|d)?\d*)(?:\s+(.*))?$`)
)

func ParseTransaction(input string) (domain.Transaction, error) {
	trimmed := strings.TrimSpace(input)
	if trimmed == "" {
		return domain.Transaction{}, ErrNoTransaction
	}
	if runeLen(trimmed) > MaxInputRunes {
		return domain.Transaction{}, fmt.Errorf("%w: input too long", ErrNoTransaction)
	}

	txType := domain.TransactionExpense
	text := trimmed
	if withoutPrefix, ok := trimIncomePrefix(text); ok {
		txType = domain.TransactionIncome
		text = strings.TrimSpace(withoutPrefix)
	}
	if text == "" {
		return domain.Transaction{}, ErrNoTransaction
	}

	match := transactionRe.FindStringSubmatch(text)
	if match == nil {
		return domain.Transaction{}, ErrNoTransaction
	}
	category := normalizeText(match[1])
	amountToken := normalizeText(match[2])
	note := ""
	if len(match) > 3 {
		note = normalizeText(match[3])
	}
	if category == "" {
		return domain.Transaction{}, ErrNoTransaction
	}
	if runeLen(category) > MaxCategoryRunes {
		return domain.Transaction{}, fmt.Errorf("%w: category too long", ErrNoTransaction)
	}
	if runeLen(note) > MaxNoteRunes {
		return domain.Transaction{}, fmt.Errorf("%w: note too long", ErrNoTransaction)
	}
	amount, err := ParseAmount(amountToken)
	if err != nil {
		return domain.Transaction{}, err
	}
	tx := domain.Transaction{Category: capitalizeFirst(category), Amount: amount, Note: note, Type: txType}
	if err := tx.Validate(); err != nil {
		return domain.Transaction{}, err
	}
	return tx, nil
}

func trimIncomePrefix(input string) (string, bool) {
	normalized := normalizeForIntent(input)
	prefixes := []string{"thu nhap", "nhan", "thu"}
	for _, prefix := range prefixes {
		if normalized == prefix {
			return "", true
		}
		if strings.HasPrefix(normalized, prefix+" ") {
			return trimLeadingWords(input, len(strings.Fields(prefix))), true
		}
	}
	return input, false
}

func trimLeadingWords(input string, count int) string {
	input = strings.TrimSpace(input)
	for count > 0 && input != "" {
		fields := strings.Fields(input)
		if len(fields) <= 1 {
			return ""
		}
		first := fields[0]
		idx := strings.Index(input, first) + len(first)
		input = strings.TrimSpace(input[idx:])
		count--
	}
	return input
}

func normalizeText(input string) string {
	return strings.Join(strings.Fields(strings.TrimSpace(input)), " ")
}

func capitalizeFirst(input string) string {
	input = strings.TrimSpace(input)
	if input == "" {
		return ""
	}
	r, size := utf8.DecodeRuneInString(input)
	if r == utf8.RuneError && size == 0 {
		return input
	}
	return string(unicode.ToUpper(r)) + input[size:]
}

func runeLen(s string) int { return len([]rune(s)) }

File: src/parser/transaction.ts
import {
  type Transaction,
  TRANSACTION_EXPENSE,
  TRANSACTION_INCOME,
  validateTransaction,
} from '../domain/transaction.ts';
import { normalizeForIntent } from './intent.ts';
import { parseAmount } from './amount.ts';

export const MAX_INPUT_RUNES = 2_000;
export const MAX_CATEGORY_RUNES = 120;
export const MAX_NOTE_RUNES = 500;

export class TransactionNotRecognizedError extends Error {
  override name = 'TransactionNotRecognizedError';

  constructor(message = 'transaction not recognized') {
    super(message);
  }
}

const TRANSACTION_PATTERN = /^(.+?)\s+([0-9][0-9.,]*\s*(?:k|tr|m|đ|d)?\d*)(?:\s+(.*))?$/u;

export function parseTransaction(input: string): Transaction {
  const trimmed = input.trim();
  if (!trimmed) throw new TransactionNotRecognizedError();
  if (runeLength(trimmed) > MAX_INPUT_RUNES) {
    throw new TransactionNotRecognizedError('transaction not recognized: input too long');
  }

  let type = TRANSACTION_EXPENSE;
  let text = trimmed;
  const incomePrefix = trimIncomePrefix(text);
  if (incomePrefix !== undefined) {
    type = TRANSACTION_INCOME;
    text = incomePrefix.trim();
  }
  if (!text) throw new TransactionNotRecognizedError();

  const match = TRANSACTION_PATTERN.exec(text);
  if (!match) throw new TransactionNotRecognizedError();
  const category = normalizeText(match[1]!);
  const amountToken = normalizeText(match[2]!);
  const note = match[3] === undefined ? '' : normalizeText(match[3]);
  if (!category) throw new TransactionNotRecognizedError();
  if (runeLength(category) > MAX_CATEGORY_RUNES) {
    throw new TransactionNotRecognizedError('transaction not recognized: category too long');
  }
  if (runeLength(note) > MAX_NOTE_RUNES) {
    throw new TransactionNotRecognizedError('transaction not recognized: note too long');
  }

  const transaction: Transaction = {
    category: capitalizeFirst(category),
    amount: parseAmount(amountToken),
    note,
    type,
  };
  validateTransaction(transaction);
  return transaction;
}

function trimIncomePrefix(input: string): string | undefined {
  const normalized = normalizeForIntent(input);
  for (const prefix of ['thu nhap', 'nhan', 'thu']) {
    if (normalized === prefix) return '';
    if (normalized.startsWith(`${prefix} `)) {
      return trimLeadingWords(input, prefix.split(' ').length);
    }
  }
  return undefined;
}

function trimLeadingWords(input: string, count: number): string {
  let value = input.trim();
  while (count > 0 && value) {
    const match = /^\S+/u.exec(value);
    if (!match) return '';
    value = value.slice(match[0].length).trim();
    count--;
  }
  return value;
}

function normalizeText(input: string): string {
  return input.trim().split(/\s+/u).filter(Boolean).join(' ');
}

function capitalizeFirst(input: string): string {
  const characters = Array.from(input.trim());
  if (characters.length === 0) return '';
  characters[0] = characters[0]!.toLocaleUpperCase('vi-VN');
  return characters.join('');
}

function runeLength(value: string): number {
  return Array.from(value).length;
}
